classlib/src/main/java/org/teavm/classlib/java/math/TBigInteger.java

```diff
@@ -1097,6 +1146,76 @@ public TBigInteger pow(int exp) {
         return TMultiplication.pow(this, exp);
     }
 
+    /**
```

Comment:
Just curious where you've taken this?

Reply:
The overall idea is taken from chapter 11 *Integer Square Root* using Newton's Method in the book [Hacker's Delight](https://doc.lagout.org/security/Hackers%20Delight.pdf), combined with some random code samples I've seen over the internet.

If I remember correctly, I got this reference from a doc-comment when scrolling through the OpenJDK classlib. Given that I have *seen* their code before writing this implementation, you might consider this implementation to be tainted by their GPL'ed code, even if not copied per-se.

If you prefer a "cleaner" implementation in this regard, I can provide an algorithmic write-up as a base for someone who hasn't seen their code.

Reply:
Yeah, I understand. This is fine by me.
Copyright and licensing is not always clear enough if you are not a lawyer and these instances are usually only clearly decided by a judge (with horrendous sums of money thrown at lawyers most people don't have). :shrug: